File: csharp/src/SecurityShield.cs
using System.Text.Json;

namespace B4N1Web;

/// <summary>
/// SecurityShield provides URL security validation with caching
/// </summary>
public class SecurityShield
{
    private readonly string _dbPath;
    private readonly int _cacheDays;
    private readonly Dictionary<string, CacheEntry> _cache;

    private class CacheEntry
    {
        public bool IsSafe { get; set; }
        public DateTime Expires { get; set; }
    }

    public SecurityShield(string? dbPath = null, int cacheDays = 7)
    {
        _dbPath = dbPath ?? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".b4n1web/security.db");
        _cacheDays = cacheDays;
        _cache = new Dictionary<string, CacheEntry>();
    }

    /// <summary>
    /// Extract domain from URL
    /// </summary>
    private string? ExtractDomain(string url)
    {
        try
        {
            if (Uri.TryCreate(url, UriKind.Absolute, out var uri))
            {
                return uri.Host.ToLower();
            }
        }
        catch
        {
        }
        return null;
    }

    /// <summary>
    /// Check if URL is safe to navigate
    /// </summary>
    public (bool isSafe, bool needsApiCheck) IsUrlSafe(string url)
    {
        var domain = ExtractDomain(url);
        if (domain == null)
        {
            return (true, false);
        }

        if (!_cache.TryGetValue(domain, out var entry))
        {
            return (true, true);
        }

        if (DateTime.Now > entry.Expires)
        {
            _cache.Remove(domain);
            return (true, true);
        }

        return (entry.IsSafe, false);
    }

    /// <summary>
    /// Mark a domain as safe or unsafe
    /// </summary>
    public void MarkDomain(string domain, bool isSafe)
    {
        var normalized = domain.ToLower();
        _cache[normalized] = new CacheEntry
        {
            IsSafe = isSafe,
            Expires = DateTime.Now.AddDays(_cacheDays)
        };
    }

    /// <summary>
    /// Clear all cached domains
    /// </summary>
    public void ClearCache()
    {
        _cache.Clear();
    }
}

/// <summary>
/// Navigation result
/// </summary>
public class NavigateResult
{
    public string Url { get; set; } = string.Empty;
    public bool Success { get; set; }
    public string? Markdown { get; set; }
    public List<string>? Links { get; set; }
    public string? Error { get; set; }
}

/// <summary>
/// Navigation helper with security check
/// </summary>
public static class NavigationHelper
{
    /// <summary>
    /// Navigate to URL with optional security check
    /// </summary>
    public static async Task<NavigateResult> NavigateAsync(
        string url,
        bool ignoreSecurity = false,
        SecurityShield? shield = null,
        BrowserOptions? browserOptions = null)
    {
        if (!ignoreSecurity)
        {
            shield ??= new SecurityShield();
            var (isSafe, _) = shield.IsUrlSafe(url);
            if (!isSafe)
            {
                return new NavigateResult
                {
                    Url = url,
                    Success = false,
                    Error = "URL flagged as unsafe by security check"
                };
            }
        }

        try
        {
            var browser = new AgentBrowser(browserOptions);
            var page = await browser.GotoAsync(url);
            browser.Close();

            return new NavigateResult
            {
                Url = page.Url,
                Success = true,
                Markdown = page.Markdown,
                Links = page.Links
            };
        }
        catch (Exception ex)
        {
            return new NavigateResult
            {
                Url = url,
                Success = false,
                Error = ex.Message
            };
        }
    }
}

File: java/src/main/java/com/b4n1/web/SecurityShield.java
package com.b4n1.web;

import java.net.URI;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * SecurityShield provides URL security validation with caching.
 *
 * Fall-safe: returns safe=true if any error occurs.
 */
public class SecurityShield {
    private final String dbPath;
    private final int cacheDays;
    private final Map<String, CacheEntry> cache;

    private static class CacheEntry {
        boolean isSafe;
        long expires;

        CacheEntry(boolean isSafe, long expires) {
            this.isSafe = isSafe;
            this.expires = expires;
        }
    }

    public SecurityShield() {
        this(null, 7);
    }

    public SecurityShield(String dbPath, int cacheDays) {
        this.dbPath = dbPath != null ? dbPath 
            : System.getProperty("user.home") + "/.b4n1web/security.db";
        this.cacheDays = cacheDays;
        this.cache = new HashMap<>();
    }

    /**
     * Extract domain from URL.
     */
    private String extractDomain(String url) {
        try {
            URI uri = new URI(url);
            return uri.getHost().toLowerCase();
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Check if URL is safe to navigate.
     */
    public SecurityCheckResult isUrlSafe(String url) {
        String domain = extractDomain(url);
        if (domain == null) {
            return new SecurityCheckResult(true, false);
        }

        CacheEntry entry = cache.get(domain);
        if (entry == null) {
            return new SecurityCheckResult(true, true);
        }

        if (System.currentTimeMillis() > entry.expires) {
            cache.remove(domain);
            return new SecurityCheckResult(true, true);
        }

        return new SecurityCheckResult(entry.isSafe, false);
    }

    /**
     * Mark a domain as safe or unsafe.
     */
    public void markDomain(String domain, boolean isSafe) {
        String normalized = domain.toLowerCase();
        long expires = System.currentTimeMillis() + (cacheDays * 24L * 60L * 60L * 1000L);
        cache.put(normalized, new CacheEntry(isSafe, expires));
    }

    /**
     * Clear all cached domains.
     */
    public void clearCache() {
        cache.clear();
    }

    /**
     * Result of security check.
     */
    public static class SecurityCheckResult {
        public final boolean isSafe;
        public final boolean needsApiCheck;

        public SecurityCheckResult(boolean isSafe, boolean needsApiCheck) {
            this.isSafe = isSafe;
            this.needsApiCheck = needsApiCheck;
        }
    }
}
